include/circt/Dialect/OM/OMOpInterfaces.td

```diff
@@ -15,6 +15,102 @@
+  }];
+  let methods = [
+    InterfaceMethod<"Get the class-like field to type map",
+      "std::optional<mlir::Type>", "getFieldType", (ins "mlir::StringAttr":$name)>,
+    InterfaceMethod<"Print class fields like",
+      "void", "printClassFieldsLike", (ins "mlir::OpAsmPrinter &":$printer,
+                                           "llvm::SmallVector<mlir::Value>":$operands),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{ 
+        // * assumes fieldNames dictionary attr exists
+        // * assumes each name in fieldNames has an entry in fieldTypes
+        //   dictionary attr mapping the name to type
+        // * perFieldAttrs, perFieldLocs optional attr/locs for each field
+        // * each field is emitted on a separate line
+        mlir::Operation *op = $_op.getOperation();
+
+        printer << "(";
+        printer.increaseIndent();
+
+        llvm::SmallVector<mlir::StringAttr> fieldNames = $_op.getFieldNames();
+
+        mlir::ArrayRef<mlir::Attribute> perFieldAttrs;
+        if (auto attrs = op->getAttr("perFieldAttrs"))
+          perFieldAttrs = mlir::cast<mlir::ArrayAttr>(attrs).getValue();
+
+        mlir::ArrayRef<mlir::Attribute> perFieldLocs;
+        if (auto locs = op->getAttr("perFieldLocs"))
+          perFieldLocs = mlir::cast<mlir::ArrayAttr>(locs).getValue();
+
+        for (unsigned i = 0; i < fieldNames.size(); i++) {
+          mlir::StringAttr name = fieldNames[i];
+          if (i > 0) {
+            printer << ",";
+          }
+          printer.printNewline();
+
+          // print symbol
+          printer.printSymbolName(name.getValue());
+
+          // print operand if exists
+          if (!operands.empty()) {
+            printer << " ";
+            printer.printOperand(operands[i]);
+          }
+
+          // print type
+          std::optional<mlir::Type> type = $_op.getFieldType(name);
+          assert(type.has_value() && "class field has no type");
+          printer << " : ";
+          printer.printType(type.value());
+
+          // print optional attr
+          if (!perFieldAttrs.empty())
+            if (auto fieldAttr = mlir::dyn_cast<mlir::DictionaryAttr>(perFieldAttrs[i]))
+              printer.printOptionalAttrDict(fieldAttr.getValue());
+
+          // print optional loc
+          if (!perFieldLocs.empty()) {
+            mlir::Location loc = mlir::cast<mlir::Location>(perFieldLocs[i]);
+            if (!mlir::isa<mlir::UnknownLoc>(loc))
+              printer.printOptionalLocationSpecifier(loc);
+          }
+        }
+        printer.decreaseIndent();
+        if (!fieldNames.empty())
+          printer.printNewline();
+        printer << ")";
+        printer.printOptionalAttrDict(op->getAttrs(),
+                                      /*elidedAttrs=*/{"fieldTypes", "fieldNames",
+                                                       "fieldIdxs", "perFieldAttrs",
+                                                       "perFieldLocs"});
+      }]>,
```

Comment:
i think we generally want to keep these attributes as inherent attributes not discardable. Could you add these attributes to op's arguments in ODS?

Reply:
I agree, and if you haven't seen these terms, there is a brief description here: https://mlir.llvm.org/docs/LangRef/#attributes.